optuna/_gp/_search_space.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from enum import IntEnum
+import math
+from typing import Any
+
+import numpy as np
+import scipy.stats.qmc
+
+import optuna
+
+
+class ParamType(IntEnum):
+    LINEAR = 0
+    LOG = 1
+    CATEGORICAL = 2
+
+
+@dataclass(frozen=True)
+class SearchSpace:
+    param_type: np.ndarray
+    bounds: np.ndarray
+    step: np.ndarray
+
+
+def untransform_one_param(
+    param_value: np.ndarray, param_type: ParamType, bounds: tuple[float, float], step: float
+) -> np.ndarray:
+    if param_type == ParamType.CATEGORICAL:
+        return param_value
+    else:
+        bounds2 = (bounds[0] - 0.5 * step, bounds[1] + 0.5 * step)
+        if param_type == ParamType.LOG:
+            bounds2 = (math.log(bounds2[0]), math.log(bounds2[1]))
+        param_value = param_value * (bounds2[1] - bounds2[0]) + bounds2[0]
+        if param_type == ParamType.LOG:
+            param_value = np.exp(param_value)
+        return param_value
+
+
+def transform_one_param(
```

Comment:
I think normalize is more specific.
```suggestion
def normalize_one_param(
```